app/models/sac_cas/role.rb

```diff
@@ -8,23 +8,73 @@
-                    '#{date.strftime("%Y-%m-%d")}',
-                    COALESCE(DATE(roles.deleted_at), '9999-12-31'),
-                    COALESCE(DATE(roles.archived_at), '9999-12-31'),
-                    COALESCE(roles.delete_on, '9999-12-31')
-                  ),
-                  DATE(roles.created_at)
-                )
-            )/365
+              CASE
+                -- membership_years is only calculated for Mitglied roles
+                WHEN roles.type != 'Group::SektionsMitglieder::Mitglied' THEN 0
+                ELSE (
+                    TIMESTAMPDIFF(
+                        YEAR, 
+                        DATE(CONVERT_TZ(roles.created_at, "UTC", "CET")),
+                        LEAST(
+                            DATE('#{date.strftime("%Y-%m-%d")}'),
+                            -- add one day to include last day in calculation
+                            COALESCE(DATE(CONVERT_TZ(roles.deleted_at, "UTC", "CET")) + INTERVAL 1 DAY, '9999-12-31'),
+                            COALESCE(DATE(CONVERT_TZ(roles.archived_at, "UTC", "CET")) + INTERVAL 1 DAY, '9999-12-31'),
+                            COALESCE(roles.delete_on + INTERVAL 1 DAY, '9999-12-31')
+                        )
+                    ) 
+                + CASE
+                -- check if dates are the same, to not add fractional year
+                  WHEN (
+                      MONTH(LEAST(
+                          DATE('#{date.strftime("%Y-%m-%d")}'),
+                          COALESCE(DATE(CONVERT_TZ(roles.deleted_at, "UTC", "CET")) + INTERVAL 1 DAY, '9999-12-31'),
+                          COALESCE(DATE(CONVERT_TZ(roles.archived_at, "UTC", "CET")) + INTERVAL 1 DAY, '9999-12-31'),
+                          COALESCE(roles.delete_on + INTERVAL 1 DAY, '9999-12-31')
+                      )) = MONTH(CONVERT_TZ(roles.created_at, "UTC", "CET")) AND
+                      DAY(LEAST(
+                          DATE('#{date.strftime("%Y-%m-%d")}'),
+                          COALESCE(DATE(CONVERT_TZ(roles.deleted_at, "UTC", "CET")) + INTERVAL 1 DAY, '9999-12-31'),
+                          COALESCE(DATE(CONVERT_TZ(roles.archived_at, "UTC", "CET")) + INTERVAL 1 DAY, '9999-12-31'),
+                          COALESCE(roles.delete_on + INTERVAL 1 DAY, '9999-12-31')
+                      )) = DAY(CONVERT_TZ(roles.created_at, "UTC", "CET"))
+                  ) THEN 0
+                   ELSE
+                        (
+                      -- calculate the fractional year
+                      DATEDIFF(
+                        LEAST(
+                          '#{date.strftime("%Y-%m-%d")}',
+                          COALESCE(DATE(CONVERT_TZ(roles.deleted_at, "UTC", "CET")), '9999-12-31'),
+                          COALESCE(DATE(CONVERT_TZ(roles.archived_at, "UTC", "CET")), '9999-12-31'),
+                          COALESCE(roles.delete_on, '9999-12-31')
+                        ),
```

Comment:
Dieser Block wiederholt sich und könnte in eine eigene Methode extrahiert werden. Müsste bei den roles Daten nicht wie oben auch noch + 1 Day gerechnet werden? Möglicherweise nicht, in den Specs sieht es eigentlich so richtig aus...